File: apple_watch_app/PathOfNurWatch/Refresh/SnapshotFreshnessPolicy.swift
import Foundation

struct SnapshotFreshnessPolicy {
    func evaluate(
        now: Date,
        snapshot: WatchDaySnapshot?,
        lastRefreshAt: Date?,
        lastSuccessfulSyncAt: Date?,
        hasPendingActions: Bool
    ) -> SnapshotFreshnessState {
        guard let snapshot else {
            return SnapshotFreshnessState(
                snapshotDate: nil,
                lastRefreshAt: lastRefreshAt,
                lastSuccessfulSyncAt: lastSuccessfulSyncAt,
                freshnessStatus: .expired,
                reason: "Missing snapshot"
            )
        }

        let calendar = Calendar.current
        if !calendar.isDate(snapshot.date, inSameDayAs: now) {
            return SnapshotFreshnessState(
                snapshotDate: snapshot.date,
                lastRefreshAt: lastRefreshAt,
                lastSuccessfulSyncAt: lastSuccessfulSyncAt,
                freshnessStatus: .expired,
                reason: "Snapshot belongs to a different day"
            )
        }

        let age = now.timeIntervalSince(lastRefreshAt ?? snapshot.lastSyncAt ?? snapshot.date)
        if hasPendingActions {
            return SnapshotFreshnessState(
                snapshotDate: snapshot.date,
                lastRefreshAt: lastRefreshAt,
                lastSuccessfulSyncAt: lastSuccessfulSyncAt,
                freshnessStatus: age > 60 * 60 ? .stale : .aging,
                reason: "Pending local actions"
            )
        }
        if age <= 15 * 60 {
            return SnapshotFreshnessState(snapshotDate: snapshot.date, lastRefreshAt: lastRefreshAt, lastSuccessfulSyncAt: lastSuccessfulSyncAt, freshnessStatus: .fresh, reason: "Recent refresh")
        }
        if age <= 60 * 60 {
            return SnapshotFreshnessState(snapshotDate: snapshot.date, lastRefreshAt: lastRefreshAt, lastSuccessfulSyncAt: lastSuccessfulSyncAt, freshnessStatus: .aging, reason: "Refresh advisable soon")
        }
        return SnapshotFreshnessState(snapshotDate: snapshot.date, lastRefreshAt: lastRefreshAt, lastSuccessfulSyncAt: lastSuccessfulSyncAt, freshnessStatus: .stale, reason: "Snapshot is aging")
    }
}

File: wear_os_app/app/src/main/java/com/pathofnur/watch/refresh/SnapshotFreshnessPolicy.kt
package com.pathofnur.watch.refresh

import com.pathofnur.watch.domain.WatchDaySnapshot
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

class SnapshotFreshnessPolicy {
    fun evaluate(
        now: Instant,
        snapshot: WatchDaySnapshot?,
        lastRefreshAt: Long?,
        lastSuccessfulSyncAt: Long?,
        hasPendingActions: Boolean
    ): SnapshotFreshnessState {
        if (snapshot == null) {
            return SnapshotFreshnessState(null, lastRefreshAt, lastSuccessfulSyncAt, SnapshotFreshnessStatus.Expired, "Missing snapshot")
        }
        val today = LocalDate.now(ZoneId.systemDefault()).toString()
        if (snapshot.date != today) {
            return SnapshotFreshnessState(snapshot.date, lastRefreshAt, lastSuccessfulSyncAt, SnapshotFreshnessStatus.Expired, "Snapshot belongs to a different day")
        }
        val ageMillis = now.toEpochMilli() - (lastRefreshAt ?: lastSuccessfulSyncAt ?: now.toEpochMilli())
        if (hasPendingActions) {
            return SnapshotFreshnessState(snapshot.date, lastRefreshAt, lastSuccessfulSyncAt, if (ageMillis > 60 * 60 * 1000) SnapshotFreshnessStatus.Stale else SnapshotFreshnessStatus.Aging, "Pending local actions")
        }
        return when {
            ageMillis <= 15 * 60 * 1000 -> SnapshotFreshnessState(snapshot.date, lastRefreshAt, lastSuccessfulSyncAt, SnapshotFreshnessStatus.Fresh, "Recent refresh")
            ageMillis <= 60 * 60 * 1000 -> SnapshotFreshnessState(snapshot.date, lastRefreshAt, lastSuccessfulSyncAt, SnapshotFreshnessStatus.Aging, "Refresh advisable soon")
            else -> SnapshotFreshnessState(snapshot.date, lastRefreshAt, lastSuccessfulSyncAt, SnapshotFreshnessStatus.Stale, "Snapshot is aging")
        }
    }
}

